examples/bsc-p2p/src/block_import/service.rs

```diff
@@ -0,0 +1,266 @@
+
+            match engine.new_payload(payload).await {
+                Ok(payload_status) => match payload_status.status {
+                    PayloadStatusEnum::Valid => {
+                        // Send forkchoice update if the payload is valid
+                        if let Err(err) =
+                            Self::send_fork_choice_update(&engine, &provider, hash, number).await
+                        {
+                            return Outcome::<T> { peer: peer_id, result: Err(err) };
+                        }
+
+                        Outcome::<T> {
+                            peer: peer_id,
+                            result: Ok(BlockValidation::ValidBlock { block }),
+                        }
+                    }
+                    PayloadStatusEnum::Invalid { validation_error } => Outcome::<T> {
+                        peer: peer_id,
+                        result: Err(BlockImportError::Other(validation_error.into())),
+                    },
+                    PayloadStatusEnum::Syncing => Outcome::<T> {
+                        peer: peer_id,
+                        result: Err(BlockImportError::Other(
+                            PayloadStatusEnum::Syncing.as_str().into(),
+                        )),
+                    },
+                    _ => Outcome::<T> {
+                        peer: peer_id,
+                        result: Err(BlockImportError::Other("Unsupported payload status".into())),
+                    },
+                },
+                Err(err) => {
+                    Outcome::<T> { peer: peer_id, result: Err(BlockImportError::Other(err.into())) }
+                }
+            }
+        })
+    }
+
+    /// Add a new block import task to the pending imports
+    fn on_new_block(&mut self, block: BlockMsg<T>, peer_id: PeerId) {
+        self.pending_imports.push(self.import(block, peer_id));
+    }
+}
+
+impl<Provider: BlockNumReader + Clone + 'static + Unpin, T: EngineTypes> Future
+    for ImportService<Provider, T>
+{
+    type Output = Result<(), Box<dyn std::error::Error>>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.get_mut();
+
+        // Receive new blocks from network
+        while let Poll::Ready(Some((block, peer_id))) = this.from_network.poll_recv(cx) {
+            this.on_new_block(block, peer_id);
+        }
+
+        // Process completed imports and send outcomes to network
+        while let Poll::Ready(Some(outcome)) = this.pending_imports.poll_next_unpin(cx) {
+            if let Err(e) = this.to_network.send(outcome) {
+                return Poll::Ready(Err(Box::new(e)));
+            }
+        }
+
+        Poll::Pending
+    }
+}
+
+/// Determines the head block hash according to Parlia consensus rules:
+/// 1. Follow the highest block number
+/// 2. For same height blocks, pick the one with lower hash
+pub(crate) fn canonical_head(
```

Comment:
we can likely start with a simple struct ParliaConsesnsus struct and include all of the logic in there, this should make upcoming refactoring work easier